# HCI 574 - lecture 36 - Web application with Flask 

April 15, 2024

(Post lecture change: fixed gte_links() so it now also accepts https URLs)

- reminder to fill out the 2. lecture feedback assignment  

#### Recap
- HTML as a container for text formatting and GUI scripting in static web pages
- run a local server to serve up a HTML doc we crated from scratch
- used HTTP (via the requests module) to download data and HTML docs 
- used the BeautyfulSoup module to download a HTML file and parse (analyze) it to pull out all its links (hrefs)
- scraped files from a web site 

<p>

### Today
- we'll write a Web application using the [Flask Microframework](http://flask.pocoo.org/), ([documentation](http://flask.pocoo.org/docs/)) 
- for now deploy the app on a local Flask server
- The web application will:  
    - Use a simple javascript GUI to give the server a URL to a web page (input) 
    - Have the server pull out all links from that web page (processing with BeautifulSoup from last lecture) 
    - Display the links

<p>

- This is based on  [Flasks Hello world with a customized greeting](https://player.oreilly.com/videos/9781491911921)   (video is part of the O'Reilly book: Flask Web Development which I have skimmed over and liked ..)
- this workflow is similar to __HW10__, except that you will have to analyze an RSS feed instead of a web page (i.e. you won't use BeautifulSoup!)

<p>

- next lecture: deploy our app to a remote server on [pythonanywhere.com](https://www.pythonanywhere.com). Please get an account (free) with them before next time.
- next-next lecture: analyzing RSS newsfeeds (XML format)

### Serve up a dynamically generated HTML page

- You will will run a Flask server on your local PC at a certain port number (I'll use 8080) 
- Last time we made our simple server load a HTML file (index.html)
- Now, we will generate this HTML "file" as a string of text and have the server serve up this dynamically generated "page" instead
- To look at that page from the running server, paste `http://127.0.0.1:8080` into your browser (the :8080 is important!)

### How to make the server serve up a specific HTML page
- Web servers typically serve up many different pages
- Every page we want the app to serve needs to have a `route` definition
- the route to /  (the last / in `http://127.0.0.1:8080/`)  is defined in Flask with:


```py
@app.route("/") 
def main_page(): <- will be called when server needs to serve up /
   ...
```
   
- main_page() is a sort of  __callback__ function
- what's the `@` ? It's a [decorator](https://realpython.com/primer-on-python-decorators/), [more](https://sumit-ghosh.com/articles/demystifying-decorators-python/). 
- Think of it of a special kind of method `route()` of the `app` class that reads in the def on the next line. 
- this case here defines a route method for our app that's called main_page() and that runs automatically when the browser requests the root page / of our server running at `http://127.0.0.1:8080`

<br>

- the page is the same as my javascript example file from last time but it's generated as a string inside the method
- generating html code directly works but is rather crude, a better way is to use [templates](https://blog.miguelgrinberg.com/post/the-flask-mega-tutorial-part-ii-templates)), which we'll do next

In [2]:
%pip install requests
%pip install bs4
%pip install Flask

Note: you may need to restart the kernel to use updated packages.Requirement already satisfied: requests in c:\users\david\appdata\local\packages\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\localcache\local-packages\python310\site-packages (2.28.1)




[notice] A new release of pip is available: 23.0.1 -> 24.0
[notice] To update, run: C:\Users\david\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.0.1 -> 24.0
[notice] To update, run: C:\Users\david\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


     ---------------------------------------- 0.0/101.7 kB ? eta -:--:--
     -------------- ---------------------- 41.0/101.7 kB 653.6 kB/s eta 0:00:01
     -------------------------------------- 101.7/101.7 kB 1.5 MB/s eta 0:00:00
     ---------------------------------------- 0.0/133.2 kB ? eta -:--:--
     -------------------------------------  133.1/133.2 kB 4.0 MB/s eta 0:00:01
     -------------------------------------- 133.2/133.2 kB 2.7 MB/s eta 0:00:00
     ---------------------------------------- 0.0/226.8 kB ? eta -:--:--
     ------------------------------------- 226.8/226.8 kB 13.5 MB/s eta 0:00:00
     ---------------------------------------- 0.0/97.9 kB ? eta -:--:--
     ---------------------------------------- 97.9/97.9 kB ? eta 0:00:00
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.0.1 -> 24.0
[notice] To update, run: C:\Users\david\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


## Step 1 - generate a simple html page as a single string
- instantiate an app
- add a route for `/`
- when the route is run, make a html document (in a simple string) and return it
- the browser will display (render) that page
- once your server runs, click on http://127.0.0.1:8080/ to open a browser that requests the / (root) page
- to stop the server use _Interrupt_ 
- clicking submit will have no effect (later)

In [3]:
# Web app with flask - step 1
from flask import Flask, request    # import some classes from flask module 
#(BTW this is NOT the requests module we used last time, this is a sub-module of Flask

# define an app object
app = Flask("step1")

# for this app, define a method to respond when the root page URL (/) is requested
@app.route("/") #@ is a decorator
def main_page():
    
    # return the HTML code that the browser will receive and process  
    # here it contains a chunk of JS that it will run when you mouse over the h2 text
    # this is the same as the JS example from last lecture where it was inside a HTML file
    return """
      <html>
        <body>
          
          <!-- JavaScript -->
          <script>
              function mOver(obj)
              {
                  obj.innerHTML=Date()
              }
              function mOut(obj)
              {
                  obj.innerHTML="Hello World of WebApps - Step 1";
              }
          </script>
          <!-- JavaScript -->
          
           <h2 onmouseover="mOver(this)" onmouseout="mOut(this)" >
              Step 1 - just one page, html is generate as a simple string
          </h2>
          
          
          <!-- not yet connected, so nothing will happen when you hit the submit button -->
          <form action="">
              Example of two Radio buttons: <br>       
              <input type="radio" name="status" value="On">On<br>
              <input type="radio" name="status" value="Off" checked>Off<br>
      
              <br>Example of a drop-down list with a pre selected value: <br>     
              <select name="lang">
              <option value="C++">C++</option>
              <option value="Java">Java</option>
              <option value="Python" selected>Python</option>
              <option value="Fortran">Fortran</option>
              </select>
      
             <br><br>Example of a text entry field<br> 
                <input type="text" name="URL"   
                    value="www.iastate.edu" 
                    size="40">
             <br><br>       
             <input type="submit" value="Submit Form content">
           </form>            
      
        </body>
      </html>"""

In [4]:
app.run(debug=False, port=8080)  # host on http://127.0.0.1:8080/
# Don't use Debug=True, is seems to not release the socket and you'll get a bind error when you run it next!

# Should say:  * Running on http://127.0.0.1:8080/ (Press CTRL+C to quit)
# To stop the server when using jupyter use Stop Cell Execution or Restart

 * Serving Flask app 'step1'
 * Debug mode: off


 * Running on http://127.0.0.1:8080
Press CTRL+C to quit
127.0.0.1 - - [16/Apr/2024 23:08:03] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [16/Apr/2024 23:08:04] "GET /favicon.ico HTTP/1.1" 404 -
127.0.0.1 - - [16/Apr/2024 23:08:08] "GET /?status=Off&lang=Java&URL=www.iastate.edu HTTP/1.1" 200 -


## Step 2: Use a template file instead of a string to serve the html page 
-  `step2.html` in the template folder is our template file
- by default all template files are expected to be in the template (sub) folder
- it's the same as my html string in step 1, except for `{{ title }}`

```HTML
<html>
    <body>
          
      <script>
      ...
      	function mOut(obj)
		  {
			  obj.innerHTML="{{ title }}";
		  }
	  </script>
      
	   <h2 onmouseover="mOver(this)" onmouseout="mOut(this)" >
		  {{ title }}
	  </h2>
	  
	  
	  <form action="">
      ...
      </form>             

    </body>
</html>
```
	
- `{{ title }}` with the curly brackets is NOT html indicates a __template variable__
- Flask uses [jinja2](https://codeburst.io/jinja-2-explained-in-5-minutes-88548486834e), yet another special language, to allow us to fill specific "special" parts of a template with __our__ values from Python.
- `render_template()` will do this variable inlining and then give the template text (with our variable) to the browser to render (display)
- Here, we just fill `{{ title }}` with our own text by giving render_template the argument `title` (which matches the word inside the `{{ }}`) with the string we want `{{ title }}` to be replaced with.

<p>

- `render_template('step2.html', title="Step 2 - using a template")` will paste `Step 2 - using a template` into any `{{ title }}` within the html file

### Running the step 2 server 

- make sure the old server has stopped, then run the cell below and go to http://127.0.0.1:8080/ If it hangs hit Restart
- as debug, I print out templated html doc  
- You can see that the code is much more concise b/c I don't have to create html as a string, I just fill a few spots in the index.html template with my own settings and give the filled file to the browser
- You will see that the h2 tag is now filled with the title value I used in `render_template()` (Step 2 - using a template)
- If you hit the Submit button and then look at the server log, you'll see that it transmits the set values to our server: `127.0.0.1 - - [10/Apr/2020 13:02:56] "GET /?status=Off&lang=Python&URL=www.iastate.edu HTTP/1.1" 200 -`
- In step 3, we'll learn how to actually catch them!

In [5]:
from flask import Flask, request, render_template # now also import the render template class
app = Flask("step2")

@app.route("/")  
def main_page():
    html_str = render_template('step2.html', title="Step 2 - using a template file") # title will be inlined in {{ title }}
    print(html_str) # DEBUG
    return html_str  # give it to the browser to display the inline page

app.run(debug=False, port=8080) 

# http://127.0.0.1:8080/

 * Serving Flask app 'step2'
 * Debug mode: off


 * Running on http://127.0.0.1:8080
Press CTRL+C to quit
127.0.0.1 - - [16/Apr/2024 23:09:17] "GET / HTTP/1.1" 200 -


<html>
    <body>
          
	  <!-- JavaScript -->
	  <script>
		  function mOver(obj)
		  {
			  obj.innerHTML=Date()
		  }
		  function mOut(obj)
		  {
			  obj.innerHTML="Step 2 - using a template file";
		  }
	  </script>
	  <!-- JavaScript -->
	  
	   <h2 onmouseover="mOver(this)" onmouseout="mOut(this)" >
		  Step 2 - using a template file
	  </h2>
	  
	  
	  <!-- not yet connected to server, so nothing will happen when you hit the submit button for now ... -->
	  <form action="">
		  Example of two Radio buttons: <br>       
		  <input type="radio" name="status" value="On">On<br>
		  <input type="radio" name="status" value="Off" checked>Off<br>
  
		  <br>Example of a drop-down list with a pre selected value: <br>     
		  <select name="lang">
		  <option value="C++">C++</option>
		  <option value="Java">Java</option>
		  <option value="Python" selected>Python</option>
		  <option value="Fortran">Fortran</option>
		  </select>
  
		 <br><br>Example of a text entry field<br> 
		

## Step 3: Submit data to the server
- the template for step 3 (step3.html) is this (same as step 2 but without the mouse callback stuff and the radio buttons)
  
```HTML
<html>
  <body>
    
    {{ title }} <br><br>  
  	
    <form action="/result/" method="get">   
		
        Select a value:
		<select name="lang">
	  		<option value="C++">C++</option>
	  		<option value="Java">Java</option>
	  		<option value="Python" selected>Python</option>
	  		<option value="Fortran">Fortran</option>
		</select>
	 	
        Enter a URL: 
		<input  type="text"  name="URL"
			value="www.iastate.edu" 
			size="40">
		
        <input type="submit" value="Submit URL to Server">
    
    </form>             
  
  </body>
</html>
```

- `form` contains 2 data fields, which belong to the 2 GUI widgets:
    - `lang` contains a value selected via a _pull down list_ (default is `Python`)
    - `URL` contains a string given via a _text field_ (default value is `www.iastate.edu`)
- the value of these fields is transmitted to the server via `<input type="submit">`, which is a special button widget

<p>

- Once submit is pressed, the value of our  widgets inside the `form` tag are transmitted to the server:
- the `/result/` in `<form action="/result/" method="get">` means that these values are to be sent to a __new__ page
- Using the "get" method will transmit values to the server by inlining the inside the URL as query parameters, which is everything after the ?. You can see that in the actual URL and also in the server log, 
- ex: `"GET /?lang=Python&URL=www.iastate.edu HTTP/1.1"` encodes the value for `lang` as `Python` and the value for `URL` as `www.iastate.edu`
- Effectively the submit button calls this URL: `http://127.0.0.1:8080/result/?lang=Python&URL=www.iastate.edu`

<p>

- the new page is called `result` (i.e. `http://127.0.0.1:8080/result/` instead of the main page at `http://127.0.0.1:8080/`
- We still will need the main route (cell below):

In [6]:
from flask import Flask, request, render_template  

app = Flask("step3")

# Main route as our landing page
@app.route("/") 
def main_page():
    # fill in values to template
    html_str = render_template('step3.html', title="Step 3 - submitting values from HTML form")
    return html_str  # give it to the browser

#### the /result/ page
- but - we need to also set up a new route for `/result/`, which the server will request and give the values to
- as this route will be called with query parameters (`/result/?lang=Python&URL=www.iastate.edu`), flask automatically creates a __request__ object that contains the data from the form

<p>

- the request object has a args dictionary, which contains the name of the data field (lang or URL) and its value
- e.g. ` request.args["URL"]` has the value value given to the parameter `URL`, i.e. `www.iastate.edu`

<p>

- once we have the values, I create a super simple html doc (as string) that: 
    -  shows the names and their values as a simple paragraph (`<p>` tag)
    - has another `get` form with a submit button that jumps back to / (root):
    ```HTML
        <form action="/" method="get">
           <input type="submit" value="Back to Main page"> 
        </form>
    ```

<p>

- after defining the result route, we can again start the app.
- Make sure the step 2 server is stopped (interrupt), then run the cell below and point your browser again to http://127.0.0.1:8080/
- If that doesn't work (happened to me!) just __Restart the Kernel__

<p>

- Finally, I added a "Back to Main page" submit button that will call the root (`/`) ... but without any query parameters 
- because of that this sets all fields in the main back to the page default values. 
- Alternatively, you can hit your browser's back arrow, for which the values are cached

In [7]:
from flask import Flask, request, render_template  

app = Flask("step3")

# Main route as our landing page
@app.route("/") 
def main_page():
    html_str = render_template('step3.html', title="Step 3 - submitting values from HTML form")
    return html_str  # give it to the browser

# route for grabbing the browers data (HTML form)
@app.route('/result/', methods=["GET"])
def result_page():
    
    # request.args is a dict with the name and the value of the select and 
    print("URL:", request.args["URL"])   # value for name="URL"  part in form
    
    html = """
        <html>
          <body>
            <p>Server here! I just got this URL:  """ + request.args["URL"] + """ </p>
            
            <form action="/" method="get">
              <input type="submit" value="Back to Main page"> 
            </form>
          
          </body>
        </html>"""
    return html

app.run(debug=False, port=8080)

# http://127.0.0.1:8080/

 * Serving Flask app 'step3'
 * Debug mode: off


 * Running on http://127.0.0.1:8080
Press CTRL+C to quit
127.0.0.1 - - [17/Apr/2024 00:07:29] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [17/Apr/2024 00:07:33] "GET /result/?language=Python&URL=www.iastate.edu/index HTTP/1.1" 200 -


URL: www.iastate.edu/index


127.0.0.1 - - [17/Apr/2024 00:07:37] "GET /result/?language=Python&URL=www.iastate.edu/index HTTP/1.1" 200 -


URL: www.iastate.edu/index


### Class exercise: also print out the selected language
- add code that also "catches" the string coming from the UI element with the name "language" in the current (step3) template
- use that string to "print out" the selected language inside its own paragraph

In [ ]:
from flask import Flask, request, render_template  

app = Flask("step3")

# Main route as our landing page
@app.route("/") 
def main_page():
    html_str = render_template('step3.html', title="Step 3 - submitting values from HTML form")
    return html_str  # give it to the browser

# route for grabbing the browers data (HTML form)
@app.route('/result/', methods=["GET"])
def result_page():
    
    # request.args is a dict with the name and the value of the select and 
    print("Language:", ?????     )
    print("URL:", request.args["URL"])   # value for name="URL"  part in form
    
    html = """
        <html>
          <body>
            <p>Server here! I just got this URL:  """ + request.args["URL"] + """ </p>
            <p>Language is: """ + ??? + """</p>
            <form action="/" method="get">
              <input type="submit" value="Back to Main page"> 
            </form>
          
          </body>
        </html>"""
    return html

app.run(debug=False, port=8080)

# http://127.0.0.1:8080/

In [8]:
'''




















'''
# Solution
from flask import Flask, request, render_template  

app = Flask("step3")

# Main route as our landing page
@app.route("/") 
def main_page():
    html_str = render_template('step3.html', title="Step 3 - submitting values from HTML form")
    return html_str  # give it to the browser

# route for grabbing the browers data (HTML form)
@app.route('/result/', methods=["GET"])
def result_page():
    
    # request.args is a dict with the name and the value of the select and 
    print("Language:", request.args["language"])
    print("URL:", request.args["URL"])   # value for name="URL"  part in form
    
    html = """
        <html>
          <body>
            <p>Server here! I just got this URL:  """ + request.args["URL"] + """ </p>
            <p>Language is: """ + request.args["language"] + """</p>
            <form action="/" method="get">
              <input type="submit" value="Back to Main page"> 
            </form>
          
          </body>
        </html>"""
    return html

app.run(debug=False, port=8080)

# http://127.0.0.1:8080/

 * Serving Flask app 'step3'
 * Debug mode: off


 * Running on http://127.0.0.1:8080
Press CTRL+C to quit
127.0.0.1 - - [17/Apr/2024 00:10:03] "GET / HTTP/1.1" 200 -


## Step 4: Analyze (parse) the URL with BeautifulSoup and display its links
- For the last step, we will parse the URL give to us via BeautifulSoup 
- I will re-use the parsing code from last time, where we created a HTML doc containing all links scraped from that URL
- For this, all I need is a text input widget for the URL and the submit button on the main page, so no more language selection

<p>

- The step4.html is:
```
<html>
 <body>
   {{ title }}<br><br>  
   <form action="/result/" method="get">  
	 Give me a URL to scrape for links: 
	 <input type="text"  name="URL"
			value="www.iastate.edu" 
			size="40">
	<input type="submit" value="Commence Scraping!">
   </form>             
 </body>
</html>
```

<p>

- on submit, the server gets the URL and again transmits it to result as: `/result/?URL=www.iastate.edu`
- The first part is the same as for step 3, execpt I also import BeautifulSoup for later ...

In [ ]:
from flask import Flask, request, render_template   
import requests # this is the requests (with s!) module
import bs4 # BeautifulSoup parser for html

app = Flask("step4")
@app.route("/") 
def main_page():
    html_str = render_template('step4.html', title="Step 4 - get a URL, scrape for links and display them")
    return html_str  # give it to the browser 


#### Extract the data
- in the `/result/` route we will again get the URL from the `request.arg` dict
- we also need fix up the URL so it starts with `http://` and ends with `/` otherwise it's not a proper, valid URL

<p>

- Using `requests` (nothing to do with the request object from before!) we load the html text of that URL and extract all anchor tags via BeautifulSoup. The href value of each anchor tag contains the URL we are looking for.
- The function `get_links()` will return a list of all (non-local) link URLs (href) from that list of anchor tags

In [ ]:
def get_links(anchors_tags):
    links = []
    for a in anchors_tags:
        l = a.get('href')
        if l != None and l.startswith("http"): # only collect external (internet) links
            links.append(str(l))    # convert to string and append to list
    return links

#### Make a web page to show the scraped data (/result/ route)
- make a long string with all link URLs, each inside a anchor tag
- make a html doc from it with a back to Main page button

In [ ]:

@app.route('/result/', methods=["GET"])
def result_page():
    
    # request.args is a dict with the name and the value of the select and 
    url = request.args["URL"]  # value for name = "URL"  in input part of form
    print("Downloading page from", url)
    
    # if url doesn't start with http://, put it in front, otherwise requests.get() gets confused
    if not url.startswith("http://") and not url.startswith("https://"):
        url = "https://" + url
    
    # if url doesn't end with / add that
    if url[-1] != "/": 
        url += '/'
    
    
    # GET webpage from the url, 
    wp = requests.get(url)  
    wp.raise_for_status()  # raises an exception if there was a problem when getting the url
    
    # create a BeautifulSoup html parser object 
    # wp.text contains the web page text we want to analyse
    soup = bs4.BeautifulSoup(wp.text, "html.parser")  
    
    # pull out all link URLs  (the <a> or anchor tag)
    anchor_tags = soup.find_all('a')
    links = get_links(anchor_tags)
    print ("found", len(links), "links in", url)
    
    # make a long string with all URLs wrapped into anchor tags
    link_URLs_string = ""
    for l in links:
        link_URLs_string = link_URLs_string + "<a href=\"" + l + "\"> " + l + " </a><br>"
    
    # wrap HTML body around the links and add a back to main button   
    html = """
        <html>
          <body>
            These links were found in """ + url + """
            <form action="/" method="get">
              <input type="submit" value="Back to Main page"> 
            </form>""" + link_URLs_string + """
          </body>
        </html>"""
    return html # show finished web page

#### Error handling
- You may want to protect your server code from simply stopping when an error occurs
- Example: user enters an invalid URL to scrape
- You can define an error handler for each type of error as defined by the error code (e.g. 404)
- here I use 500 instead of 404 with seems to be a more general error
- Again, this is done in flask by defining a function (`errorhandler()`) via a decorator 
- `error` contains the error text (very technical!)
- I just show the invalid webpage for this error as a simple string

In [ ]:
@app.errorhandler(500)
def page_not_found(error):
    print(error) 
    s = "Error with " + str(request.args["URL"]) + "<br>" + str(error)   
    s = s + "<br>Hit the Back button and try another site to scrape"
    return s

#### Step 4 code in full - using a .py file
- I've copied the cell below into step4.py
- Load step4.py and let's step through it first 

In [ ]:
# Step 4 code in one single cell
from flask import Flask, request, render_template   
import requests # this is the requests (with s!) module
import bs4 # BeautifulSoup parser for html

def get_links(anchors_tags):
    links = []
    for a in anchors_tags:
        l = a.get('href')
        if l != None and (l[:5] == "http:" or l[:6] == "https"): # only collect external (internet) links
            links.append(str(l))    # convert to string and append to list
    return links

app = Flask("step4")
@app.route("/") 
def main_page():
    html_str = render_template('step4.html', title="Step 4 - get a URL, scrape for links and display them")
    return html_str  # give it to the browser 

@app.route('/result/', methods=["GET"])
def result_page():
    
    # request.args is a dict with the name and the value of the select and 
    url = request.args["URL"]  # value for name = "URL"  in input part of form
    print("Downloading page from", url)
    
    # if url doesn't start with http://, put it in front, otherwise requests.get() gets confused
    if not url.startswith("http://") and not url.startswith("https://"):
        url = "https://" + url
    
    # if url doesn't end with / add that
    if url[-1] != "/": 
        url += '/'
    
    # GET webpage from the url, 
    wp = requests.get(url)  
    wp.raise_for_status()  # raises an exception if there was a problem when getting the url
    
    # create a BeautifulSoup html parser object 
    # wp.text contains the web page text we want to analyse
    soup = bs4.BeautifulSoup(wp.text, "html.parser")  
    
    # pull out all link URLs  (the <a> or anchor tag)
    anchor_tags = soup.find_all('a')
    links = get_links(anchor_tags)
    print ("found", len(links), "links in", url)
    
    # make a long string with all URLs wrapped into anchor tags
    link_URLs_string = ""
    for l in links:
        link_URLs_string = link_URLs_string + "<a href=\"" + l + "\"> " + l + " </a><br>"
    
    # wrap HTML body around the links and add a back to main button   
    html = """
        <html>
          <body>
            These links were found in """ + url + """
            <form action="/" method="get">
              <input type="submit" value="Back to Main page"> 
            </form>""" + link_URLs_string + """
          </body>
        </html>"""
    return html # show finished web page

@app.errorhandler(500)
def page_not_found(error):
    print(error) 
    s = "Error with " + str(request.args["URL"]) + "<br>" + str(error)   
    s = s + "<br>Hit the Back button and try something else ...)"
    return s

app.run(debug=False, port=8080)

# http://127.0.0.1:8080/

### For next lecture:
- sign up for a [beginner](https://www.pythonanywhere.com/pricing/) (=free) account at https://pythonanywhere.com
- you will get an email to confirm your account
- pythonanywhere.com let's you run python code on their server. However, in the free mode, you  can't run jupyter, you're limited to putting code into a .py file (stored on their server) and running that file via the command line or just using a python shell
- next lecture we will walk through how to deploy this flask app to their server 